### PR TITLE
Reintroduce animated input for Markdown composer

### PR DESCRIPTION
**Pull request by robertKozik**
### Details
It's the follow up PR for enabling live-markdown-preview text input (introduced in https://github.com/Expensify/App/pull/34292). In the original PR MarkdownTextInput component was used directly, which disabled pre-erasing of text input in the worklet. This PR creates new wrapper AnimatedMarkdownTextInput.

I cannot use the RNTextInput (which is used before) as we don't need MarkdownTextInput in every place where later wrapper is used.

### Fixed Issues

$ https://github.com/Expensify/App/issues/27977
PROPOSAL:

### Tests

1. Log in
2. Write and send message
3. Verify that message is being send correctly
- [X] Verify that no errors appear in the JS console

### Offline tests

Same as Test steps
### QA Steps

Same as Test steps
- [X] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android: Native
    - [X] Android: mWeb Chrome
    - [X] iOS: Native
    - [X] iOS: mWeb Safari
    - [X] MacOS: Chrome / Safari
    - [X] MacOS: Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [X] If the PR modifies the form input styles:
    - [X] I verified that all the inputs inside a form are aligned with each other.
    - [X] I added `Design` label so the design team can review the changes.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/61146594/fc320144-b3de-4ab8-8a53-c6c5774c9897

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/61146594/00d3cd98-b4b2-4db5-a8e4-d5d58193e1fe

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/61146594/fd781ff0-65d4-43f4-b8e9-48ac931d7f0c

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/61146594/b989a4c2-dc13-49cb-b1ff-97869c7f9146

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/61146594/e0d3abb5-fdd4-4cbe-8270-75c52e9743e8

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/61146594/b16d2b0c-403f-4781-a4dd-cf2e9e7612c7

</details>

### CODE REVIEW

File: src/components/RNMarkdownTextInput.tsx [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+                }
+                ref(refHandle as AnimatedTextInputRef);
+            }}
+            // eslint-disable-next-line
+            {...props}
+        />
+    );
+}
+
+RNMarkdownTextInput.displayName = 'RNTextInputWithRef';
```
Comment (tomekzaw) [inline src/components/RNMarkdownTextInput.tsx:34]:
```suggestion
RNMarkdownTextInputWithRef.displayName = 'RNMarkdownTextInputWithRef';
```

File: src/components/RNMarkdownTextInput.tsx [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import type {MarkdownTextInputProps} from '@expensify/react-native-live-markdown';
+import {MarkdownTextInput} from '@expensify/react-native-live-markdown';
+import type {ForwardedRef} from 'react';
+import React from 'react';
+import type {TextInput} from 'react-native';
+import Animated from 'react-native-reanimated';
+import useTheme from '@hooks/useTheme';
+
+// Convert the underlying TextInput into an Animated component so that we can take an animated ref and pass it to a worklet
+const AnimatedTextInput = Animated.createAnimatedComponent(MarkdownTextInput);
```
Comment (tomekzaw) [inline src/components/RNMarkdownTextInput.tsx:10]:
```suggestion
const AnimatedMarkdownTextInput = Animated.createAnimatedComponent(MarkdownTextInput);
```

File: src/components/RNMarkdownTextInput.tsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+type AnimatedTextInputRef = typeof AnimatedTextInput & TextInput & HTMLInputElement;
```
Comment (tomekzaw) [inline src/components/RNMarkdownTextInput.tsx:12]:
```suggestion
type AnimatedMarkdownTextInputRef = typeof AnimatedMarkdownTextInput & TextInput & HTMLInputElement;
```

File: src/components/RNMarkdownTextInput.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+function RNMarkdownTextInput(props: MarkdownTextInputProps, ref: ForwardedRef<AnimatedTextInputRef>) {
```
Comment (tomekzaw) [inline src/components/RNMarkdownTextInput.tsx:14]:
```suggestion
function RNMarkdownTextInputWithRef(props: MarkdownTextInputProps, ref: ForwardedRef<AnimatedMarkdownTextInputRef>) {
```

File: src/components/RNMarkdownTextInput.tsx [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    const theme = useTheme();
+
+    return (
+        <AnimatedTextInput
```
Comment (tomekzaw) [inline src/components/RNMarkdownTextInput.tsx:18]:
```suggestion
        <AnimatedMarkdownTextInput
```

File: src/components/RNMarkdownTextInput.tsx [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+            }}
+            // eslint-disable-next-line
+            {...props}
+        />
+    );
+}
+
+RNMarkdownTextInput.displayName = 'RNTextInputWithRef';
+
+export default React.forwardRef(RNMarkdownTextInput);
```
Comment (tomekzaw) [inline src/components/RNMarkdownTextInput.tsx:36]:
```suggestion
export default React.forwardRef(RNMarkdownTextInputWithRef);
```

**Review by tomekzaw — COMMENTED** [review]
Just a few suggestions to make it consistent with [RNTextInput.tsx](https://github.com/Expensify/App/blob/main/src/components/RNTextInput.tsx)

**Comment by tomekzaw** [comment]
Fixes #35787.

File: src/components/Composer/index.native.tsx [hunk 7]
```diff
@@ -1,9 +1,9 @@
-import {MarkdownTextInput} from '@expensify/react-native-live-markdown';
 import type {ForwardedRef} from 'react';
 import React, {useCallback, useEffect, useMemo, useRef} from 'react';
 import type {TextInput} from 'react-native';
 import {StyleSheet} from 'react-native';
-import type {AnimatedTextInputRef} from '@components/RNTextInput';
+import type {AnimatedMarkdownTextInputRef} from '@components/RNMarkdownTextInput';
+import MarkdownTextInput from '@components/RNMarkdownTextInput';
```
Comment (tomekzaw) [inline src/components/Composer/index.native.tsx:6]:
```suggestion
import RNMarkdownTextInput from '@components/RNMarkdownTextInput';
```

**Review by tomekzaw — COMMENTED** [review]
Just one more suggestion

**Comment by robertKozik** [comment]
Hi @thesahindia When can we expect your review?

**Review by thesahindia — APPROVED** [review]
LGTM!

**Review by thienlnam — APPROVED** [review]

**Review by thienlnam — APPROVED** [review]

**Pull request merged**